src/paths.md

```diff
@@ -1,72 +1,137 @@
-<span class="parenthetical">symbol (`<`)</span>. In type expression context, the
-final namespace qualifier is omitted.
+## Types of paths
 
-Two examples of paths with type arguments:
+### Simple Paths
+
+> **<sup>Syntax</sup>**\
+> _SimplePath_ :\
+> &nbsp;&nbsp; `::`<sup>?</sup> _SimplePathSegment_ (`::` _SimplePathSegment_)<sup>\*</sup>
+>
+> _SimplePathSegment_ :\
+> &nbsp;&nbsp; [IDENTIFIER] | `super` | `self` | `$crate`
+
+Simple paths are used in [visibility] markers, [attributes], [macros], and [`use`] items.
+Examples:
 
 ```rust
-# struct HashMap<K, V>(K,V);
-# fn f() {
-# fn id<T>(t: T) -> T { t }
-type T = HashMap<i32,String>; // Type arguments used in a type expression
-let x  = id::<i32>(10);       // Type arguments used in a call expression
-# }
+use std::io::{self, Write};
+mod m {
+    #[clippy::cyclomatic_complexity = "0"]
+    pub (in super) fn f1() {}
+}
 ```
 
-Paths can be denoted with various leading qualifiers to change the meaning of
-how it is resolved:
+### Paths in expressions
+
+> **<sup>Syntax</sup>**\
+> _PathInExpression_ :\
+> &nbsp;&nbsp; `::`<sup>?</sup> _PathExprSegment_ (`::` _PathExprSegment_)<sup>\*</sup>
+>
+> _PathExprSegment_ :\
+> &nbsp;&nbsp; _PathIdentSegment_ (`::` [_Generics_])<sup>?</sup>
+>
+> _PathIdentSegment_ :\
+> &nbsp;&nbsp; [IDENTIFIER] | `super` | `self` | `Self` | `$crate`
 
-* Paths starting with `::` are considered to be global paths where the
-  components of the path start being resolved from the crate root. Each
-  identifier in the path must resolve to an item.
+Paths in expressions allow for paths with generic arguments to be specified. They are
+used in various places in [expressions] and [patterns].
+
+The `::` token is required before the opening `<` for generic parameters to avoid
+ambiguity with the less-than operator. This is colloquially known as "turbofish" syntax.
 
 ```rust
-mod a {
-    pub fn foo() {}
+(0..10).collect::<Vec<_>>();
+Vec::<u8>::with_capacity(1024);
+```
+
+## Qualified paths
+
+> **<sup>Syntax</sup>**\
+> _QualifiedPathInExpression_ :\
+> &nbsp;&nbsp; `<` [_Type_] (`as` _TypePath_)? `>` (`::` _PathExprSegment_)<sup>\*</sup>
+
+Fully qualified paths allow for disambiguating the path for [trait implementations] and for
+specifying [canonical paths](#canonical-paths).
+
+```rust
+struct S;
+impl S {
+    fn f() { println!("S"); }
 }
-mod b {
-    pub fn foo() {
-        ::a::foo(); // call a's foo function
-    }
+trait T1 {
+    fn f() { println!("T1 f"); }
 }
-# fn main() {}
+impl T1 for S {}
+trait T2 {
+    fn f() { println!("T2 f"); }
+}
+impl T2 for S {}
+S::f();  // Calls the inherent impl.
+<S as T1>::f();  // Calls the T1 trait function.
+<S as T2>::f();  // Calls the T2 trait function.
+```
+
+### Paths in types
+
+> **<sup>Syntax</sup>**\
+> _TypePath_ :\
+> &nbsp;&nbsp; `::`<sup>?</sup> _TypePathSegment_ (`::` _TypePathSegment_)<sup>\*</sup>
+>
+> _TypePathSegment_ :\
+> &nbsp;&nbsp; _PathIdentSegment_ (`::`<sup>?</sup> [_Generics_])<sup>?</sup>
```

Comment:
There's also a variant that allows specifying associated items (i.e. `Iterator<Item=i32>`). I'm not sure if having two variants here is better or worse that specifying when it's allow in the prose.

Reply:
I'm not sure if I'm following.  Is that form not part of the generics (_TypeParam_ in [generics](https://github.com/rust-lang-nursery/reference/blob/master/src/items/generics.md))?

Reply:
Yes, I must have missed that

Reply:
I realized I had the wrong rule listed, updated.